Bugfix: signature verification with different query encoding

arothuis opened the pull request:
[136419577](https://www.pivotaltracker.com/story/show/136419577)

If the SP sends a SAMLRequest that is encoded differently as query parameters than PHP's standard way of encoding (`http_build_query`), signature verification should not break.

The spec states: 
> (...) [N]ote that URL-encoding is not canonical; that is, there are multiple legal encodings for a given value. The relying party MUST therefore perform the verification step using the original URL-encoded values it received on the query string. It is not sufficient to re-encode the parameters after the have been processed by software because > the resulting encoding may not match the signer's encoding.

For further reference, see: [Oasis Standard regarding SAML Bindings, 15 March 2005, page 18, line 620](https://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf)

We were [re-encoding the query string](https://github.com/SURFnet/Stepup-saml-bundle/blob/3c0371d428af4a23e98438ccd76df890cf36edbf/src/SAML2/AuthnRequest.php#L305-L319), but we should rely on the encoding given.

rjkip commented on src/SAML2/AuthnRequestFactory.php:
```diff
@@ -145,7 +146,7 @@ public static function createNewRequest(ServiceProvider $serviceProvider, Identi
      * @return SAML2_Configuration_PrivateKey|XMLSecurityKey
      */
-    private static function loadPrivateKey(SAML2_Configuration_PrivateKey $key)
+    public static function loadPrivateKey(SAML2_Configuration_PrivateKey $key)
```
Looks like you used this during testing once upon a time, but it doesn't seem like this method needs to be public.

rjkip commented on src/Tests/Component/Signing/AuthnRequestSigningTest.php:
```diff
@@ -0,0 +1,216 @@
+     * @test
+     * @group Signing
+     */
+    public function a_signature_is_verified_if_the_data_to_sign_corresponds_with_how_it_is_included_in_the_http_query_regardless_of_how_it_is_encoded(
```
`signatures_are_verified_regardless_of_encoding_used_by_sender` is terser and seems to get the message across just fine :slightly_smiling_face:

rjkip commented on src/Tests/Component/Signing/AuthnRequestSigningTest.php:
```diff
@@ -0,0 +1,216 @@
+        $authnRequestWithDefaultEncoding = $this->createSignedAuthnRequestPreparingDataToSignWith(
+            [$this, 'encodeDataToSignWithPhpsHttpBuildQuery']
+        );
+        $authnRequestWithCustomEncoding  = $this->createSignedAuthnRequestPreparingDataToSignWith(
```
I'd drop the `PreparingDataToSignWith` suffix. No need, though.

rjkip commented on src/Tests/Component/Signing/AuthnRequestSigningTest.php:
```diff
@@ -0,0 +1,216 @@
+            [$this, 'encodeDataToSignWithPhpsHttpBuildQuery']
+        );
+        $authnRequestWithCustomEncoding  = $this->createSignedAuthnRequestPreparingDataToSignWith(
+            [$this, 'encodeDataToSignWithCustomEncoder']
```
What about `encodeDataToSignWithCustomHttpQueryEncoding`? This terminology is also used in the assertion below.

rjkip commented on src/Tests/Component/Signing/AuthnRequestSigningTest.php:
```diff
@@ -0,0 +1,216 @@
+    {
+        $encodedParams = http_build_query($params);
+
+        return str_replace('%3A', ':', $encodedParams);
```
Please document why we opted to do custom encoding here.

rjkip commented on src/Tests/Component/Signing/AuthnRequestSigningTest.php:
```diff
@@ -0,0 +1,216 @@
+    public function a_signature_is_not_verified_if_parameter_order_of_the_corresponding_http_query_is_not_correct()
+    {
+        $authnRequestWithModifiedDataToSign = $this->createSignedAuthnRequestPreparingDataToSignWith(
+            [$this, 'changeParameterAndEncodeDataToSign']
```
What about `encodeDataToSignUsingIncorrectParameterOrder`?

rjkip commented on src/Tests/Component/Signing/AuthnRequestSigningTest.php:
```diff
@@ -0,0 +1,216 @@
+        );
+        $this->assertTrue($signatureWithCustomEncodingIsVerified,
+            'The signature of the authn request with a custom http query encoding '
+            . 'should correspond with its data to sign, but it does not'
```
These messages can use some love. What about:

 * `The signature of an AuthnRequest signed using data-to-sign encoded according to RFC1738 should be verifiable, but it isn't`
 * `The signature of an AuthnRequest signed using data-to-sign encoded using a custom encoding should be verifiable, but it isn't`

rjkip commented on src/Tests/Component/Signing/AuthnRequestSigningTest.php:
```diff
@@ -0,0 +1,216 @@
+     */
+    private function changeParameterAndEncodeDataToSign(array $params)
+    {
+        return http_build_query(array_reverse($params, true));
```
Nice use of `array_reverse` :satisfied:

rjkip commented on src/Tests/Component/Signing/AuthnRequestSigningTest.php:
```diff
@@ -0,0 +1,216 @@
+        $domDocument          = SAML2_DOMDocumentFactory::fromString($this->authRequestNoSubject);
+        $unsignedAuthnRequest = SAML2_AuthnRequest::fromXML($domDocument->firstChild);
+        $unsignedAuthnRequest->setSignatureKey($privateKey);
+        $unsignedAuthnRequest->setCertificates([$this->getPublicKey()]);
```
Is it necessary or idiomatic to always set the signature key and certificates when you're not interested in creating a signed AuthnRequest?

arothuis replied:
I've done it this way, because I've seen it in SAML2 tests as well.
For instance, in the [MessageTest](https://github.com/simplesamlphp/saml2/blob/67ad5af84dc3215c437ccb396a7426f9f2de35b0/tests/SAML2/MessageTest.php) and the [SignatureValidationTest](https://github.com/simplesamlphp/saml2/blob/67ad5af84dc3215c437ccb396a7426f9f2de35b0/tests/SAML2/Response/SignatureValidationTest.php).

However we do not do this in our AuthnRequest factory, so I'm a bit on the fence on whether to keep it in or not.

rjkip replied:
Your examples concern requests that are actually signed :wink: Since we do not do this in our AuthnRequestFactory, I'd prefer chucking them out :slightly_smiling_face:

arothuis commented:
Tested against development environments of RA and Self Service.

rjkip commented on src/SAML2/AuthnRequest.php:
```diff
@@ -307,15 +305,40 @@ public function buildRequestQuery()
      */
     public function getSignedRequestQuery()
```
:christmas_tree: **Ignore this message if it's not yet 2017 and you're enjoying your days off!** :christmas_tree: 

---

I think it would be good to unit-test this method and/or extract the query parameter parsing algorithm and unit-test that.

Also, I woke up this morning wondering about how we parse the query parameters in this application (is this healthy? :thinking:) in both RedirectBinding and this method, and whether there should be one canonical way of reading and verifying a message that is received through the RedirectBinding. The RedirectBinding uses Symfony\Request to extract query parameters, and this method performs manual query parameter extraction. If ever the two methods deviate regarding parameter order, this could _potentially_ introduce an attack vector where we (for example) verify the signature of one `SAMLRequest` parameter (say, the first) and then handle the request that is present in the last `SAMLRequest` query parameter. Currently both methods use the last query parameter, and ignore any previous ones, so this is _not_ the case, currently. The point is that these things can go askew without us noticing it.

Concretely:

1. We may need one canonical query parsing and reading algorithm for all SAML communication to preclude any future deviations in parameter order, encoding, etc. etc. that may unexpectedly appear between Symfony\Request and the method used below.
1. We should test that this system does not allow verification of one SAMLRequest and handling of another due to confusion of same-named query parameter.
1. We should test that our RedirectBinding does not allow duplicate parameters that are relevant to the [RedirectBinding message](https://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf).

I think point 1 is important for this PR. The other points could be addressed in separate issues.

arothuis replied:
As discussed, we can mitigate this risk by checking whether the HTTP query parameters contain only one `SAMLRequest`, `Signature` and `SigAlg` and zero or one `RelayState`. We should throw exceptions if this is not the case. To ensure this behaviour (and document it through code), tests should be added. Other parameters should be left alone as per the [SAML Bindings specification](https://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf), see r. 566-567, page 16.

DRvanR commented:
Small question: as discussed last year 😛 the AuthnRequest should be split into two classes: ReceivedAuthnRequest and a variant that can only be used when creating an AuthnRequest (can't think of a name right now). Has a tech debt issue been made somewhere regarding that? Would also help with the request parsing as then it would be easier to place the receiving, parsing and signature verification of a request in either the receiving binding or the ReceivedAuthnRequest.

arothuis commented:
Thanks for the reminder. I will add a tech debt issue for that: it would indeed be nice to more clearly separate these different representations of AuthnRequest.

arothuis commented:
@rjkip

DRvanR commented on src/Http/QueryString.php:
```diff
@@ -0,0 +1,85 @@
+use Surfnet\SamlBundle\SAML2\AuthnRequest;
+use Symfony\Component\HttpFoundation\Request;
+
+final class QueryString
```
Not sure I like this object - actually quite sure I don't. It accepts everything (even requests without a querystring, or non-GET requests) and verifies nothing (the presence of the required parameters for the getSignedRequestQuery, whether or not a query is even present etc). Also, to me, methodnames are a bit weird:
- `countParameter`, what does that count, the parameters, the occurrences of a certain parameter, or the amount of occurrences with values?)
- `getSignedRequestQuery` - it returns, if possibly available since that has not been verified, a "signable" request query, that may or may not have been signed, which is not verified

Furthermore why is the signature itself not retrieved from this object?

To me it seems there was a good idea behind it, but it doesn't really live up to expectation.

I.m.h.o. Either, use this, use it well and ensure that this is the single source of truth for the request query of the current AuthnREquest (which checks whether or not the query is structurally valid etc) as a value object, or do not use it at all.

arothuis commented:
I have removed the initial QueryString concept and replaced it with a more robust value object specifically designed for received AuthnRequest. The latter have also been introduced in recent commits. New methods on the RedirectBinding and the SignatureVerifier have been created to prevent BC breaks.

arothuis commented on src/Http/ReceivedAuthnRequestQueryString.php:
```diff
@@ -0,0 +1,291 @@
+    /**
+     * @return string
+     */
+    public function getSignedQueryString()
```
Not sure, but I think this can be removed as we only really need the signableQueryString for signing and validating signatures.

DRvanR replied:
I'd argue to opposite when looking at method names :stuck_out_tongue_winking_eye:

arothuis commented:
As discussed IRL, an extensive review of recent changes will be postponed. A BETA release will be created including these changes, signifying its instability.

DRvanR commented on src/Http/ReceivedAuthnRequestQueryString.php:
```diff
@@ -0,0 +1,292 @@
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity) Extensive validation
+     * @SuppressWarnings(PHPMD.NPathComplexity) Extensive validation
+     */
+    public static function parse($requestUri)
```
In general: choose to either allow, validate and parse Request URI's or just querystrings. 
Imho we discussed the latter (and I'd still opt for that), since the logic seems to be tooled for that, and it's just plain simpler. Also makes more sense with respect to responsibility. This object (and thus this method) should not care about a request URI, whether or not that contains a `?` etc. It should only care about a single specific type of queries. Those containing a SAMLRequest (hence also the `strlen` check, that is pointless if you parse requestURI's).

DRvanR commented on src/Http/ReceivedAuthnRequestQueryString.php:
```diff
@@ -0,0 +1,292 @@
+            ));
+        }
+
+        if (strpos($requestUri, '?') === false) {
```
if it's a querystring, it should be the first character, or optional first character which is removed with an `ltrim`. Makes validation easier.

DRvanR commented on src/Http/ReceivedAuthnRequestQueryString.php:
```diff
@@ -0,0 +1,292 @@
+        if (isset($parameters[self::PARAMETER_RELAY_STATE])) {
+            $parsedQueryString->relayState = $parameters[self::PARAMETER_RELAY_STATE];
+        }
+        if (isset($parameters[self::PARAMETER_SIGNATURE])) {
```
please add newlines between two `if` statements (at multiple places)

DRvanR replied:
In general logic can be simplified to:
```php
if ($signature) {
    if (!$sigalg) { throw new Exception }

    // parse signature & sigalg
}

if ($sigalg) { throw new Exception }
```

DRvanR commented on src/Http/ReceivedAuthnRequestQueryString.php:
```diff
@@ -0,0 +1,292 @@
+            $query .= '&' . self::PARAMETER_RELAY_STATE . '=' . $this->relayState;
+        }
+
+        if ($this->isSigned()) {
```
if the query string was signed (`getSignedQueryString`), this must be true no? Thus if it is not true, this should bork.

DRvanR commented on src/Http/ReceivedAuthnRequestQueryString.php:
```diff
@@ -0,0 +1,292 @@
+
+        if ($this->isSigned()) {
+            $query .= '&' . self::PARAMETER_SIGNATURE . '=' . $this->signature
+               . '&' . self::PARAMETER_SIGNATURE_ALGORITHM . '=' . $this->signatureAlgorithm;
```
Why append the signature? That is not signed itself :wink:

DRvanR commented on src/Http/ReceivedAuthnRequestQueryString.php:
```diff
@@ -0,0 +1,292 @@
+    /**
+     * @return string
+     */
+    public function getSignableQueryString()
```
This method can go, there should not be a reason to look at querystring simply for signing purpose (i.e. that would be the to-be-created `ConstructedAuthnRequest`)

DRvanR commented on src/Http/RedirectBinding.php:
```diff
@@ -149,14 +150,118 @@ public function processSignedRequest(Request $request)
+     * @param Request $request
+     * @return ReceivedAuthnRequest
+     */
+    public function receiveUnsignedAuthnRequestFrom(Request $request)
```
Not sure this method is needed?

DRvanR replied:
And if needed, a `receiveAuthnRequest` with a simple `if ($query->isSigned())` on whether or not to validate the authnrequest should also be provided for those who want to do both? I mean, you can't know as outside user of this binding whether or not a request is signed without parsing it, which would kinda be against the purpose of this exercise. Arguably, that should be the only method, with two optional configuration parameters: `$allowSignedRequests = true` and `$requestMustBeSigned = true`.

arothuis replied:
As discussed, I will create a tech debt issue for this (and for the way AuthnRequests are built and used), so we can clean both concepts up a bit and release it in a new major.

DRvanR commented on src/SAML2/AuthnRequest.php:
```diff
@@ -307,15 +305,40 @@ public function buildRequestQuery()
     public function getSignedRequestQuery()
     {
-        $queryParams = [self::PARAMETER_REQUEST => $this->rawRequest];
+        $queryParamPairs = explode('&', $this->rawHttpQuery);
```
Are all the changes to this class still needed with the new methodology? I vaguely remember just deprecating these urging users to use the new versions instead?

DRvanR commented on src/Signing/SignatureVerifier.php:
```diff
@@ -48,6 +49,54 @@ public function __construct(KeyLoader $keyLoader, LoggerInterface $logger)
         $this->logger = $logger;
     }
 
+    public function verify(ReceivedAuthnRequestQueryString $query, ServiceProvider $serviceProvider)
```
`verifyIsSignedBy`?
